JumpscaleLibs/sal/zosv2/Zosv2.py

```diff
@@ -0,0 +1,197 @@
+from Jumpscale import j
+import netaddr
+import ipaddress
+import socket
+from nacl import public
+from nacl.encoding import Base64Encoder
+class NetworkConfig:
+    def __init__(self, private_key=None,public_key=None,wg_port=None, public_ip=None):
+        self.private_key = private_key
+        self.public_key = public_key
+        self.wg_port = wg_port
+        self.public_ip = public_ip
+
+class Zosv2(j.baseclasses.object):
+    __jslocation__ = "j.sal.zosv2"
+
+    def _init(self, **kwargs):
+        #j.clients.threebot.explorer_addr_set("explorer.testnet.grid.tf")
+        self.explorer = j.clients.threebot.explorer
+        self.config = NetworkConfig()
+
+    # consider a node up if it has received update during the last 10 minute
+    def is_up(self, node):
+        ago = j.data.time.epoch - (60 * 10)
+        return node.updated > ago
+
+    def get_node(self,ip_version):
+        #return node and public ip
+        nodes = self.explorer.actors_all.nodes.list().nodes
+        return self.find_node_public(nodes,ip_version)
+
+    def find_node_public(self, nodes, version=6):
+        # search a node that has a public ipv4 address
+        for node in filter(self.is_up, nodes):
+            for iface in node.ifaces:
+                for addr in iface.addrs:
+                    ip = ipaddress.ip_interface(addr).ip
+                    if ip.version != version:
+                        continue
+                    if ip.is_global:
+                        self.config.public_ip = str(ip)
+                        return (node, str(ip))
+        if version==4:
+            for node in filter(self.is_up, nodes):
+                ip = ipaddress.ip_interface(node.public_config.ipv4).ip
+                if ip.is_global:
+                    self.config.public_ip = str(ip)
+                    return (node, str(ip))
+            # need to check in ifaces
+    def find_free_wg_port(self, node):
+        ports = set(list(range(6000, 9000)))
+        used = set(node.wg_ports)
+        free = ports - used
+        return free.pop()
+
+    def new_reservation(self, user):
+        reservation_model = j.data.schema.get_from_url("tfgrid.workloads.reservation.1")
+        reservation = reservation_model.new()
+        reservation.customer_tid = user.id
```

Comment:
I makes more sense to set this field when you sign the reservation. 
You also need to document what is the expected type of user (well this all class lack any sort of documentation...)